app/models/deploy_metrics.rb

```diff
@@ -0,0 +1,49 @@
+# frozen_string_literal: true
+
+# Used to collect metrics of cycle time(elapsed time needed to deliver a project to production)
+# Check out https://about.gitlab.com/2019/03/04/reduce-cycle-time/ for more details on cycle time
+# The main objective for this is to gain more insight on the efficiency of the end-to-end project development process
+# There are 2 key metrics:
+# 1. PR - Production: this will show the time needed for a pull request to get deployed to production environment
+# 2. Staging - Production: this will show the time needed to deliver from staging(non-production environment)
+#                          to production environment.
+class DeployMetrics
+  NOT_APPLICABLE = 0
+
+  def initialize(deploy)
+    @deploy = deploy
+  end
+
+  def cycle_time
+    if @deploy&.production && @deploy&.status == "succeeded"
+      {pr_production: pr_production, staging_production: staging_production}
+    else
+      {pr_production: NOT_APPLICABLE, staging_production: NOT_APPLICABLE}
+    end
+  end
+
+  private
+
+  # Calculated by averaging deployment end time minus pull requests' creation time
+  def pr_production
+    prs = @deploy.changeset.pull_requests || []
+    return NOT_APPLICABLE if prs.empty?
+
+    prs.sum { |pr| @deploy.updated_at.to_i - pr.created_at.to_i } / prs.size
+  end
+
+  # Time it took from first staging deploy to finish until the first production deploy finished
+  def staging_production
+    jobs = Job.where(commit: @deploy.commit).pluck(:id)
+    stages = Stage.where(project: @deploy.project)
+
+    staging = stages.reject(&:production?)
+    production = stages - staging
+
+    first_staging_deploy = Deploy.where(job: jobs, stage: staging).reorder(:id).first
```

Comment:
- needs to only count success
 - might as well match on the commit since we have to join either way

```Ruby
scope = Deploy.successful.where(jobs: {commit: @deploy.commit}).reorder(:id)
first_staging_deploy = scope.where(stage: staging).first
first_production_deploy = scope.where(stage: production).first
```

Reply:
just `first_production_deploy = scope.where(stage: production).first` is enough maybe ? since if production deployment is unsuccessful this line will never be reached